### PR TITLE
use hadoop 2.8.0-palantir2

### PR DESCRIPTION
**Pull request by sjrand**
@robert3005 @pwoody @ash211 for real this time

List of differences between upstream branch-2.8.0 and 2.8.0-palantir1: https://github.com/palantir/hadoop/blob/branch-2.8.0/PALANTIR-CHANGELOG.md

### CODE REVIEW

**Comment by sjrand** [comment]
In addition to the tests that ran as part of the build (https://circleci.com/gh/palantir/hadoop/156), I also ran the s3a tests on my laptop (pointed to an actual s3 bucket).

The only failures were that you can overwrite a directory, which is expected based on having reverted HADOOP-13188. (The revert is hopefully temporary, but causes us to have the same behavior as we already have in 2.7.3, so not a regression.)

**Comment by sjrand** [comment]
Grr. Need to pull in `hadoop-hdfs` JAR to deal with

```
Caused by: java.lang.ClassNotFoundException: Class org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider not found
  at org.apache.hadoop.conf.Configuration.getClassByName(Configuration.java:2122)
  at org.apache.hadoop.conf.Configuration.getClass(Configuration.java:2214)
```

**Comment by sjrand** [comment]
Tried briefly to fix HDFS-11431, but it's a bit icky -- the `ConfiguredFailoverProxyProvider` class calls a bunch of other code from `hadoop-hdfs`, so it's not trivial to bring it into `hadoop-hdfs-client`. For now just added a commit to bring in `hadoop-hdfs`, which is not a regression relative to 2.7.3

**Comment by sjrand** [comment]
Apparently I've now broken a bunch of hive tests. Can't tell whether this is the cause, but looks relevant:

```
WARN org.apache.spark.sql.hive.client.IsolatedClientLoader: Failed to resolve Hadoop artifacts for the version 2.8.0-palantir1. We will change the hadoop version from 2.8.0-palantir1 to 2.4.0 and try again. Hadoop classes will not be shared between Spark and Hive metastore client. It is recommended to set jars used by Hive metastore client through spark.sql.hive.metastore.jars in the production environment.
```

**Comment by sjrand** [comment]
Hive test failures are evidently caused by inability to resolve the Hadoop version.

```
2017-02-19 22:05:56.12 - stderr> 		[FAILED     ] org.apache.hadoop#hadoop-client;2.8.0-palantir1!hadoop-client.jar:  (0ms)
2017-02-19 22:05:56.12 - stderr> 
2017-02-19 22:05:56.12 - stderr> 	==== local-m2-cache: tried
2017-02-19 22:05:56.12 - stderr> 
2017-02-19 22:05:56.12 - stderr> 	  file:/home/ubuntu/spark/dummy/.m2/repository/org/apache/hadoop/hadoop-client/2.8.0-palantir1/hadoop-client-2.8.0-palantir1.jar
2017-02-19 22:05:56.12 - stderr> 
2017-02-19 22:05:56.12 - stderr> 	==== local-ivy-cache: tried
2017-02-19 22:05:56.12 - stderr> 
2017-02-19 22:05:56.12 - stderr> 	  /home/ubuntu/.ivy2/local/org.apache.hadoop/hadoop-client/2.8.0-palantir1/jars/hadoop-client.jar
2017-02-19 22:05:56.12 - stderr> 
2017-02-19 22:05:56.12 - stderr> 	==== central: tried
2017-02-19 22:05:56.12 - stderr> 
2017-02-19 22:05:56.121 - stderr> 	  https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-client/2.8.0-palantir1/hadoop-client-2.8.0-palantir1.jar
2017-02-19 22:05:56.121 - stderr> 
2017-02-19 22:05:56.121 - stderr> 	==== spark-packages: tried
2017-02-19 22:05:56.121 - stderr> 
2017-02-19 22:05:56.121 - stderr> 	  http://dl.bintray.com/spark-packages/maven/org/apache/hadoop/hadoop-client/2.8.0-palantir1/hadoop-client-2.8.0-palantir1.jar
2017-02-19 22:05:56.121 - stderr> 
2017-02-19 22:05:56.121 - stderr> 	==== repo-1: tried
2017-02-19 22:05:56.121 - stderr> 
2017-02-19 22:05:56.121 - stderr> 	  http://www.datanucleus.org/downloads/maven2/org/apache/hadoop/hadoop-client/2.8.0-palantir1/hadoop-client-2.8.0-palantir1.jar
```

**Comment by robert3005** [comment]
Why the penultimate isn't a hit?

**Comment by robert3005** [comment]
You need to add our bintray. Won't accept with failing tests

**Comment by sjrand** [comment]
Yep, fair enough. Trying to find where to add it -- I'm a little confused because we already include `http://dl.bintray.com/palantir/releases` in our pom

**Comment by sjrand** [comment]
Woo, finally got a green build. I'm running another build now with the hadoop version bumped from 2.8.0-palantir1 to 2.8.0-palantir2, which picks up some upstream s3a fixes. If that build also passes, are you guys down to merge this?

File: core/pom.xml [hunk 1]
```diff
@@ -59,6 +59,10 @@
       <artifactId>hadoop-client</artifactId>
     </dependency>
     <dependency>
+      <groupId>org.apache.hadoop</groupId>
```
Comment (robert3005) [inline core/pom.xml]: Why is this necessary now?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 2]
```diff
@@ -113,7 +113,7 @@ private[hive] object IsolatedClientLoader extends Logging {
       SparkSubmitUtils.resolveMavenCoordinates(
         hiveArtifacts.mkString(","),
         SparkSubmitUtils.buildIvySettings(
-          Some("http://www.datanucleus.org/downloads/maven2"),
+          Some("http://dl.bintray.com/palantir/releases"),
```
Comment (robert3005) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Can this be a list? (not sure what's the exact signature). Fine if not

File: pom.xml [hunk 3]
```diff
@@ -880,6 +880,79 @@
           </exclusion>
         </exclusions>
       </dependency>
+      <!-- TODO (srand) Remove this when https://issues.apache.org/jira/browse/HDFS-11431 is fixed -->
```
Comment (robert3005) [inline pom.xml]: Can you create an issue as well and reference it here

**Comment by robert3005** [comment]
couple of pom nits, otherwise 👍

File: core/pom.xml (hunk 1 above)
Comment (sjrand) [inline core/pom.xml]: If I don't have this dependency then the resulting dist doesn't have `hadoop-hdfs-2.8.0-palantir2.jar` in it `jars/` dir.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala (hunk 2 above)
Comment (sjrand) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Apparently yes: `@param remoteRepos Comma-delimited string of remote repositories other than maven central` Will fix

File: pom.xml (hunk 3 above)
Comment (sjrand) [inline pom.xml]: https://github.com/palantir/spark/issues/111

File: core/pom.xml (hunk 1 above)
Comment (robert3005) [inline core/pom.xml]: Ok. Something has changed in the packaging since with 2.7.3 it's there. Would be good to understand what happened upstream but not blocking

**Comment by ash211** [comment]
@sjrand just to confirm, we expect a client running hadoop-2.8.0-palantir2 to work with a server of hadoop-2.7.3 ?  And what about the reverse: a client of 2.7.3 against a hadoop-palantir server?

**Comment by ash211** [comment]
Merging this PR caused this failure in the Circle test:

```
 4877 [info] ExternalSorterSuite:
 4878 [info] - empty data stream with kryo ser (79 milliseconds)
<snip>
 4888 [info] - cleanup of intermediate files in sorter (154 milliseconds)
 4889 [info] - cleanup of intermediate files in sorter with failures (141 milliseconds)
 4890 [info] - cleanup of intermediate files in shuffle (286 milliseconds)
 4891 [info] - cleanup of intermediate files in shuffle with failures *** FAILED *** (96 milliseconds)
 4892 [info]   java.lang.AssertionError: assertion failed: expected test shuffle cleanup to spill, but did not
 4893 [info]   at scala.Predef$.assert(Predef.scala:170)
 4894 [info]   at org.apache.spark.TestUtils$.assertSpilled(TestUtils.scala:178)
 4895 [info]   at org.apache.spark.util.collection.ExternalSorterSuite.org$apache$spark$util$collection$ExternalSorterSuite$$cleanupIntermediateFilesInShuffle(ExternalSorterSuite.scala:503)
 4896 [info]   at org.apache.spark.util.collection.ExternalSorterSuite$$anonfun$4.apply$mcV$sp(ExternalSorterSuite.scala:63)
 4897 [info]   at org.apache.spark.util.collection.ExternalSorterSuite$$anonfun$4.apply(ExternalSorterSuite.scala:63)
 4898 [info]   at org.apache.spark.util.collection.ExternalSorterSuite$$anonfun$4.apply(ExternalSorterSuite.scala:63)
 4899 [info]   at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
 4900 [info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
 4901 [info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
 4902 [info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
```

I think it's a flake -- https://circleci.com/gh/palantir/spark/424 -- so kicked off another build.

**Comment by sjrand** [comment]
Yes, a _Spark_ client application running 2.8.0-palantir2 against a 2.7* (or vendored equivalent) cluster has worked fine in my experience. MapReduce clients have not fared so well (classpath stuff), but that's not relevant here.

There are no plans to run palantir-hadoop on the cluster -- no reason to try to play Hadoop vendor when several other companies already do a way better job of it than I could.

File: core/pom.xml (hunk 1 above)
Comment (sjrand) [inline core/pom.xml]: I think it's https://issues.apache.org/jira/browse/HDFS-6200. Previously hadoop-client would bring in hadoop-hdfs, but after that change, hadoop-client brings in hadoop-hdfs-client instead. But then because of https://issues.apache.org/jira/browse/HDFS-11431, I have to manually add hadoop-hdfs back in.
